ksqldb-engine/src/main/java/io/confluent/ksql/internal/ThroughputMetricsReporter.java

```diff
@@ -0,0 +1,260 @@
+        this.metricRegistry = (Metrics) requireNonNull(
+            configMap.get(KsqlConfig.KSQL_INTERNAL_METRICS_CONFIG)
+        );
+        customTags.putAll(KsqlConfig.getStringAsMap(KsqlConfig.KSQL_CUSTOM_METRICS_TAGS, configMap));
+    }
+
+    @Override
+    public void metricChange(final KafkaMetric metric) {
+        if (!THROUGHPUT_METRIC_NAMES.contains(metric.metricName().name())) {
+            return;
+        }
+        addMetric(
+            metric,
+            getQueryId(metric),
+            getTopic(metric)
+        );
+    }
+
+    @Override
+    public void metricRemoval(final KafkaMetric metric) {
+        if (!THROUGHPUT_METRIC_NAMES.contains(metric.metricName().name())) {
+            return;
+        }
+
+        removeMetric(
+            metric,
+            getQueryId(metric),
+            getTopic(metric)
+        );
+    }
+
+    @Override
+    public void close() {
+    }
+
+    @Override
+    public Set<String> reconfigurableConfigs() {
+        return null;
+    }
+
+    @Override
+    public void validateReconfiguration(final Map<String, ?> configs) throws ConfigException {
+
+    }
+
+    @Override
+    public void reconfigure(final Map<String, ?> configs) {
+
+    }
+
+    @Override
+    public void contextChange(final MetricsContext metricsContext) {
+
+    }
+
+    private synchronized void addMetric(
+        final KafkaMetric metric,
+        final String queryId,
+        final String topic
+    ) {
+        final MetricName metricName = getMetricNameWithQueryIdTag(queryId, metric.metricName());
+        final Map<String, String> queryTopicThroughputMetricTags =
+            getQueryMetricTags(queryId, metric.metricName().tags());
+
+        LOGGER.debug("Adding {} metric for query={} and topic={}", metricName.name(), queryId, topic);
+
+        if (!registeredMetrics.containsKey(queryId)) {
+            registeredMetrics.put(queryId, new HashMap<>());
+        }
+        if (!registeredMetrics.get(queryId).containsKey(topic)) {
+            registeredMetrics.get(queryId).put(topic, new HashMap<>());
+        }
+
+        final ThroughputTotalMetric newMetric;
+        if (registeredMetrics.get(queryId).get(topic).containsKey(metricName)) {
+            newMetric = registeredMetrics.get(queryId).get(topic).get(metricName);
+            newMetric.add(metricName, metric);
+        } else {
+            newMetric = registeredMetrics.get(queryId).get(topic).get(metricName);
+            new ThroughputTotalMetric(
+                metricRegistry.metricName(
+                    metricName.name(),
+                    StreamsMetricsImpl.TOPIC_LEVEL_GROUP,
+                    queryTopicThroughputMetricTags),
+                metric);
+        }
+        
+        registeredMetrics.get(queryId).get(topic).put(metricName, newMetric);
+
+        metricRegistry.addMetric(
+            metricName,
+            (config, now) -> registeredMetrics.get(queryId).get(topic).get(metricName).getValue()
```

Comment:
Ultimately we just take the metrics coming out of Streams and then register new metrics which are (a) tagged with the queryId, and (b) aggregated (summed) over all values for that particular throughput metric at that topic from that query.

Essentially we're adding up the records/bytes count across all the different tasks, and across the sink nodes within those tasks if there happen to be more than one producing to a given topic (possible within Streams, not sure if you can actually do this in ksql but that doesn't matter here)